From 954d483d6f4395cd144edd124d8b456479be9e99 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Thu, 09 Jun 2022 15:34:07 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: xbar sram rings repair

This patch added support for BIRA, BISR and BIST operations
for XBAR SRAM rings.

TODO: Add Xbar memory init after bist to avoid ecc.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: I6a032d0f94ae5d09e34103b3b295249d4876206b
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 3a7e4c4..eab19f6 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -35,6 +35,8 @@
 #define DELAY_17_USEC			17
 #define DELAY_8K_USEC			8000
 #define DELAY_57_USEC			57
+#define DELAY_4K_USEC			4000
+#define DELAY_20_USEC			20
 
 #define GAUDI2_DFX_NUM			5
 #define GAUDI2_MAX_RING_NUM		28
@@ -370,6 +372,46 @@
 	}
 }
 
+static void gaudi2_xbar_mid_ring_pattern(enum gaudi2_sram_self_repair_op op)
+{
+	uint32_t expected_ones, expected_zeros;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_4K_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_20_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_4K_USEC);
+	} else {
+		preboot_puts("Invalid sram repair operation\n");
+		return;
+	}
+	regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIST)
+		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
+	else
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000000000000000001010000000;
+	expected_zeros = 0b00000000000000000000000101110000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL10, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 0b00000000000000000000000000000000;
+		expected_zeros = 0b00000000000000000000000000010000;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	}
+}
+
 static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
 {
 	uint32_t ring_value = BIT(8) | (ring << 3) | 0x5;
@@ -411,7 +453,7 @@
 				[SRAM_RING_TYPE_TPC] = gaudi2_tpc_ring_pattern,
 				[SRAM_RING_TYPE_SOB_ROT] = gaudi2_sob_rot_ring_pattern,
 				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
-				[SRAM_RING_TYPE_XBAR_MID] = NULL,
+				[SRAM_RING_TYPE_XBAR_MID] = gaudi2_xbar_mid_ring_pattern,
 				[SRAM_RING_TYPE_VDEC] = NULL,
 				[SRAM_RING_TYPE_SOB_XDMA] = NULL,
 				[SRAM_RING_TYPE_NIC] = NULL,
@@ -434,7 +476,8 @@
 			ring_type = gauid2_ring_type_table[dfx][ring];
 			if (ring_type == SRAM_RING_TYPE_SRAM ||
 				ring_type == SRAM_RING_TYPE_TPC ||
-				ring_type == SRAM_RING_TYPE_SOB_ROT) {
+				ring_type == SRAM_RING_TYPE_SOB_ROT ||
+				ring_type == SRAM_RING_TYPE_XBAR_MID) {
 				gaudi2_sram_id_sel();
 				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
 				gaudi2_sram_region_ring_pattern[ring_type](op);
@@ -552,6 +595,9 @@
 /* TODO: initialize internal tpc srams which
  * are not connected to reset block
  */
+
+/* TODO: Add Xbar memory init after bist to avoid ecc. */
+
 static void gaudi2_reset_engines(void)
 {
 	u32 tpc_iso_val;
